### PR TITLE
Support CUB prefix sum & product

### PR DESCRIPTION
**Pull request by leofang**
In the demo below, CUB gets about a 4x speed-up on a P100:
```python
>>> import cupy as cp
>>> import cupyx
>>> 
>>> a = cp.random.random(1000000)
>>>
>>> cp.cuda.cub_enabled  =False
>>> print(cupyx.time.repeat(cp.cumprod, (a, )))
cumprod             :   428.775 us   +/-57.870 (min:  396.988 / max:  976.234) us    471.108 us   +/-45.524 (min:  449.888 / max:  992.736) us
>>> cp.cuda.cub_enabled = True
>>> print(cupyx.time.repeat(cp.cumprod, (a, )))
cumprod             :   103.053 us   +/-19.460 (min:   84.242 / max:  229.079) us    132.358 us   +/-17.695 (min:  113.888 / max:  250.048) us
>>>
>>> cp.cuda.cub_enabled = False
>>> print(cupyx.time.repeat(cp.cumsum, (a, )))
cumsum              :   428.253 us   +/-55.802 (min:  398.817 / max:  978.976) us    470.163 us   +/-43.695 (min:  450.464 / max:  995.872) us
>>> cp.cuda.cub_enabled = True
>>> print(cupyx.time.repeat(cp.cumsum, (a, )))
cumsum              :   103.998 us   +/-17.858 (min:   86.243 / max:  234.290) us    133.392 us   +/-16.486 (min:  115.808 / max:  256.320) us
```

TODO:
- [ ] compare performance with `cupy.core._routines_math.scan()` (Rel: #2905)
- [ ] add a CUB call path to `cupy.core._routines_math.scan()`
- [x] check if CUB handles `NaN`s correctly
- [ ] add tests (**Question:** to here or to #2598?)

The first two action items are for speeding up certain fancy indexing cases.

### CODE REVIEW

**Comment by grlee77** [comment]
Nice. I tried this out and the only problem I encountered was in the shape of the returned result when `axis=None` and the input is nD. (The output should be a ravelled array)

Simple example with CUB enabled/disabled (the disabled behavior matches NumPy):

```Python
import cupy as cp
import numpy as np

a = cp.testing.shaped_arange((4, 5), cp, float)
cp.cumsum(a)
# array([[  1.,   3.,   6.,  10.,  15.],
#        [ 21.,  28.,  36.,  45.,  55.],
#        [ 66.,  78.,  91., 105., 120.],
#        [136., 153., 171., 190., 210.]])

cp.cuda.cub_enabled = False
cp.cumsum(a)
# array([  1.,   3.,   6.,  10.,  15.,  21.,  28.,  36.,  45.,  55.,  66.,
#         78.,  91., 105., 120., 136., 153., 171., 190., 210.])

**Comment by leofang** [comment]
Oops, thanks @grlee77! I read the spec wrong...

**Comment by emcastillo** [comment]
I was actually working on this :D.
I let you take care of it better :).
One of the things I saw is that CUB scanning does not support axis selection, so implementing that correctly may be a bit trickier.

**Comment by leofang** [comment]
Oh sorry for duplicate work @emcastillo. Which branch is it? I could cherry pick your code.
> One of the things I saw is that CUB scanning does not support axis selection, so implementing that correctly may be a bit trickier.

I don’t think CUB can support `axis != None`...

**Comment by emcastillo** [comment]
No worries, your PR is going to be better than mine so do the cub part and I will take care of improving the non-cub kernels.

**Comment by leofang** [comment]
@emcastillo This PR should be considered jointly with your #2907. Looks like `cupy.core._routines_math.scan()` is really good. It is slower than CUB only when there are >= 10^6 elements (for `float64` in this case):
```python
>>> a = cp.random.random(1000)
>>> print(cupyx.time.repeat(cp.cuda.cub.device_scan, (a, 5)))
device_scan         :    98.163 us   +/-16.937 (min:   76.599 / max:  222.032) us    106.869 us   +/-18.320 (min:   77.120 / max:  236.768) us
>>> print(cupyx.time.repeat(cp.core._routines_math._scan_for_test, (a, )))
_scan_for_test      :    67.176 us   +/- 9.349 (min:   49.020 / max:  147.861) us     76.206 us   +/-10.508 (min:   56.064 / max:  163.200) us
>>> 
>>> a = cp.random.random(100000)
>>> print(cupyx.time.repeat(cp.cuda.cub.device_scan, (a, 5)))
device_scan         :    97.013 us   +/-17.426 (min:   76.346 / max:  215.663) us    105.617 us   +/-18.885 (min:   83.296 / max:  300.320) us
>>> print(cupyx.time.repeat(cp.core._routines_math._scan_for_test, (a, )))
_scan_for_test      :    67.560 us   +/- 9.815 (min:   49.441 / max:  181.684) us     76.555 us   +/-11.008 (min:   56.416 / max:  198.016) us
>>> 
>>> a = cp.random.random(1000000)
>>> print(cupyx.time.repeat(cp.cuda.cub.device_scan, (a, 5)))
device_scan         :    94.037 us   +/-17.865 (min:   74.958 / max:  278.390) us    124.621 us   +/-16.371 (min:  106.752 / max:  301.504) us
>>> print(cupyx.time.repeat(cp.core._routines_math._scan_for_test, (a, )))
_scan_for_test      :    90.543 us   +/-17.130 (min:   70.443 / max:  201.541) us    123.582 us   +/-10.805 (min:  107.584 / max:  223.040) us
>>> 
>>> a = cp.random.random(10000000)
>>> print(cupyx.time.repeat(cp.cuda.cub.device_scan, (a, 5)))
device_scan         :    82.406 us   +/-12.235 (min:   74.969 / max:  204.807) us    650.291 us   +/- 6.245 (min:  644.096 / max:  714.784) us
>>> print(cupyx.time.repeat(cp.core._routines_math._scan_for_test, (a, )))
_scan_for_test      :    78.777 us   +/- 9.734 (min:   72.235 / max:  196.004) us    822.750 us   +/- 8.690 (min:  816.992 / max:  901.280) us
```
So, if `cupy.core._routines_math.scan()` is tweaked a bit (simply by changing the accumulator `+=` to `*=` I guess), perhaps in addition to `cumsum()` it can also do `cumprod()`, and we either wouldn't need CUB at all, or would switch to CUB only when the array is excessively large? 

(Actually I already cheated a bit in the above comparison by using `cp.cuda.cub.device_scan()` instead of `cp.cuda.cub.cub_scan()`: the latter adds about 8 us on my machine due to extra checks...)

**Comment by leofang** [comment]
> we either wouldn't need CUB at all, or would switch to CUB only when the array is excessively large?

We could also let users decide if they want to use CUB scan, as usual by toggling `cp.cuda.cub_enabled`.

**Comment by emcastillo** [comment]
I wouldn't set sub-optimal implementations by default

**Comment by leofang** [comment]
Another thing: CUB `cumsum` and `cumprod` seems to be buggy for complex numbers. I am not sure why, but this is deterministically reproducible. The cumulation stopped at the 449-th element of the result:
```python
>>> import cupy as cp
>>> a = cp.random.random(500) + 1j*cp.random.random(500)
>>> cp.cuda.cub_enabled = True
>>> b = cp.cumsum(a)  # wrong
>>> cp.cuda.cub_enabled = False
>>> c = cp.cumsum(a)  # correct
>>> cp.allclose(b[0:447],c[0:447])
array(True)
>>> cp.allclose(b[0:448],c[0:448])
array(True)
>>> cp.allclose(b[0:449],c[0:449])
array(False)
>>> b[448] == a[448]
array(True)
>>> c[448] == a[448]
array(False)
```
Looking into this...

**Comment by leofang** [comment]
> Another thing: CUB `cumsum` and `cumprod` seems to be buggy for complex numbers. 

Two observations:
1. This bug is `complex128`-only; for `complex64` arrays CUB works fine, so I suspect it's due to CUB's internal handling of data size exceeding 64 bits.
2. Adjusting the items per thread in this line
https://github.com/NVlabs/cub/blob/c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/device/dispatch/dispatch_scan.cuh#L177
will cause the bug appear in different elements (I changed from 15 to 12 and the bug started in the element No.3xx instead of 448).

@emcastillo Due to the above observations, I tend to think this is a bug in CUB (which I have not identified yet), and so I will disable CUB prefix scan for `complex128` and proceed.

**Comment by emcastillo** [comment]
I have been working on adapting the fast kernel in scan to do the batched cumsum and cumprod.
It's been a nightmare but i am pretty close to the end :D

**Comment by leofang** [comment]
Nice. We can then do a benchmark to decide if this PR is to be kept or closed.

File: cupy/cuda/cub.pyx [hunk 1]
```diff
@@ -372,8 +372,7 @@ def can_use_device_segmented_reduce(int op, x_dtype, Py_ssize_t ndim,
                                                         order)
 
 
-cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None,
-                                  bint segmented=False):
+cdef _cub_reduce_dtype_compatible(x_dtype, int op, dtype=None):
```
Comment (leofang) [inline cupy/cuda/cub.pyx]: This is to correct a mistake I made in #2562.

File: cupy/math/sumprod.py [hunk 2]
```diff
@@ -147,8 +152,15 @@ def _proc_as_batch(proc, x, axis):
...
-def _cum_core(a, axis, dtype, out, kern, batch_kern):
+def _cum_core(a, axis, dtype, out, kern, batch_kern, *, op=None):
     a = cupy.asarray(a)
+
+    if cupy.cuda.cub_enabled:
+        # result will be None if the scan is not compatible with CUB
+        result = cub.cub_scan(a, op, axis, dtype, out)
+        if result is not None:
+            return result
+
```
Comment (emcastillo) [inline cupy/math/sumprod.py]: Does this work for inplace?

**Comment by leofang** [comment]
This is a bit more complicated than I previously thought; even the original CuPy implementation might not be fully compliant. 

Because the NumPy doc is not very clear, I took a look at the NumPy's implementation `PyUFunc_GenericReduction()` and `PyUFunc_Accumulate()`, and this is my current understanding for the interplay between the `dtype` and `out` arguments in `cumsum()`/`cumprod()` (please correct me if I am wrong):
- if `dtype` is `None`:
  - if `out` is `None`:
    - `input.dtype` is used for both intermediate calculation and final output (with the caveat that integer types need to be promoted)
  - else:
    - `out.dype` is used for both intermediate calculation and final output
- else:
  - if `out` is `None`:
    - `dtype` is used for both intermediate calculation and final output
  - else:
    - `dtype` is used for intermediate calculation, and `out.dtype` is used for final output

Currently CuPy's `_cum_core` only takes care of the first branch above. @emcastillo should I follow the current CuPy behavior as is, or should I make the CUB scan fully NumPy-compliant?

I suspect we might have the same compliance issue in other reduction routines (not sure), but that should be addressed in a separate PR if any...

File: cupy/math/sumprod.py (hunk 2 above)
Comment (leofang) [reply cupy/math/sumprod.py]: It might, let me look into this.

**Comment by emcastillo** [comment]
I think that behavior should be fixed in another PR.

**Comment by leofang** [comment]
Sounds good. I'll then avoid dealing with this and let the current `_cum_core()` handle it. Then, in the next PR we just fix `_cum_core()`.

**Comment by leofang** [comment]
@emcastillo Thanks for the heads-up! I guess you are right: CUB supports in-place scan. This simplifies things quite a bit.

I verified this support in a few levels:
1. Test it directly with the new change 79780e1
2. In [Thrust's documentation](https://docs.nvidia.com/cuda/thrust/index.html#prefix-sums) in-place scans are used as an example. Since CUB is one of the Thrust's backends, it should also work.
3. By inspecting CUB's code: The `ConsumeTile()` implementation in `cub/cub/agent/agent_scan.cuh` seems to permit in-place operation.

File: cupy/math/sumprod.py (hunk 2 above)
Comment (leofang) [reply cupy/math/sumprod.py]: See https://github.com/cupy/cupy/pull/2919#issuecomment-575968811

**Comment by grlee77** [comment]
Hi @leofang. The dtype behavior you describe above (https://github.com/cupy/cupy/pull/2919#issuecomment-575705775) is also what I recently attempted to implement for cupy.mean in #2903. Seeing now that this is more common across reduction functions, then perhaps the logic should move out into a separate helper function.

**Comment by leofang** [comment]
Thanks for bringing it up, Gregory 🙂 I just took a quick look at it. While it'd be nice to have a separate helper as you said, the rules there are slightly different from https://github.com/cupy/cupy/pull/2919#issuecomment-575705775. Maybe we need a switch or flag in the helper to determine the desired behavior? 

Anyway, if you can make that helper in your PR, @emcastillo and I will be able to follow in #2907 and here. We can move the discussion there. Thanks!

**Comment by leofang** [comment]
@emcastillo I suggest we consider the PRs in the following order: #2919 (this PR) -> #2907 for two reasons.

First, I do not deal with the complexity of `out` and `dtype`, as outlined in https://github.com/cupy/cupy/pull/2919#issuecomment-575705775, and simply let the caller functions  (`cumsum` and `cumprod`) handle it, whereas in #2907 the caller code is refactored. 

Second, while Gregory's suggestion on a separate helper is tempting, I don't see immediately how such helper can be applicable since the rules vary across different functions.

If we want to make the implementation more NumPy-compliant, we can make another PR to address https://github.com/cupy/cupy/pull/2919#issuecomment-575705775 after both are merged, as you suggested earlier. 

If you agree with my suggestion, the only question I have for you is:
> - [ ] add tests (**Question:** to here or to #2598?)

Thanks 🙂

**Comment by emcastillo** [comment]
Lets do that

**Comment by leofang** [comment]
Cool, if we don’t need to add tests here, this PR is ready.

**Comment by emcastillo** [comment]
Lets add tests in #2598 thanks!

File: cupy/cuda/cub.pyx [hunk 3]
```diff
@@ -282,6 +286,37 @@ def device_csrmv(int n_rows, int n_cols, int nnz, ndarray values,
...
 
+def device_scan(ndarray x, int op):
+    cdef memory.MemoryPointer ws
+    cdef int dtype_id, x_size
+    cdef size_t ws_size
+    cdef void *x_ptr
+    cdef void *ws_ptr
+    cdef Stream_t s
+
+    if op != CUPY_CUB_CUMSUM and op != CUPY_CUB_CUMPROD:
```
Comment (emcastillo) [inline cupy/cuda/cub.pyx]: Can you rewrite this kind of checks as
`if op not in (CUPY_CUB_CUMSUM, CUPY_CUB_CUMPROD):`

Thanks

Comment (leofang) [reply cupy/cuda/cub.pyx]: No problem. This is done in 02059fb. Just curious: is this for readability or performance?

Comment (emcastillo) [reply cupy/cuda/cub.pyx]: readability

**Comment by leofang** [comment]
I fixed the compilation error. Sorry for my stupid mistake 😅

File: cupy/cuda/cub.pyx [hunk 4]
```diff
@@ -385,6 +421,23 @@ def cub_reduction(arr, op, axis=None, dtype=None, out=None, keepdims=False):
     return None
 
 
+def cub_scan(arr, op):
+    """Perform an (in-place) prefix scan using CUB.
+
+    If the specified scan is not possible, None is returned.
+    """
+    if op < CUPY_CUB_CUMSUM or op > CUPY_CUB_CUMPROD:
```
Comment (emcastillo) [inline cupy/cuda/cub.pyx]: `if op not in (CUPY_CUB_CUMSUM, CUPY_CUB_CUMPROD):`

Comment (emcastillo) [reply cupy/cuda/cub.pyx]: I will fix it later in my pr

**Comment by emcastillo** [comment]
Jenkins, test this please

File: cupy/cuda/cub.pyx (hunk 4 above)
Comment (leofang) [reply cupy/cuda/cub.pyx]: Oh, sorry for overlooking this @emcastillo 😓

**Comment by leofang** [comment]
Thanks @emcastillo!

**Pull request merged**